MonitorControl/Display.swift

```diff
@@ -140,23 +167,58 @@ class Display {
     }
 
     DispatchQueue.global(qos: .userInitiated).async {
-      guard self.ddc?.write(command: .brightness, value: UInt16(value)) == true else {
+      guard self.ddc?.write(command: .brightness, value: ddcValue) == true else {
         return
       }
 
-      self.showOsd(command: .brightness, value: value, isSmallIncrement: isSmallIncrement)
+      self.showOsd(command: .brightness, value: osdValue)
     }
 
     if let slider = brightnessSliderHandler?.slider {
-      slider.intValue = Int32(value)
+      slider.intValue = Int32(ddcValue)
+    }
+
+    self.saveValue(osdValue, for: .brightness)
+  }
+
+  func readDDCValues(for command: DDC.Command, tries _: UInt, minReplyDelay delay: UInt64?) -> (current: UInt16, max: UInt16)? {
+    var values: (UInt16, UInt16)?
+
+    if self.ddc?.supported(minReplyDelay: delay) == true {
+      os_log("Display supports DDC.", type: .debug)
+    } else {
+      os_log("Display does not support DDC.", type: .debug)
+    }
+
+    if self.ddc?.enableAppReport() == true {
+      os_log("Display supports enabling DDC application report.", type: .debug)
+    } else {
+      os_log("Display does not support enabling DDC application report.", type: .debug)
+    }
+
+    values = self.ddc?.read(command: command, tries: 10, minReplyDelay: delay)
+
+    if values != nil {
+      return values!
     }
 
-    self.saveValue(value, for: .brightness)
+    return nil
   }
 
-  func calcNewValue(for command: DDC.Command, withRel rel: Int) -> Int {
+  func calcNewValue(for command: DDC.Command, isUp: Bool, isSmallIncrement: Bool) -> Int {
     let currentValue = self.getValue(for: command)
-    return max(0, min(self.getMaxValue(for: command), currentValue + rel))
+    var nextValue = currentValue
+
+    let rel = self.stepSize(isSmallIncrement: isSmallIncrement)
+
+    if isSmallIncrement {
+      nextValue = isUp ? nextValue + rel : nextValue - rel
+    } else {
+      let largeIncrementSize = isUp ? ((nextValue + rel) % rel == 0 ? rel : rel - (nextValue + rel) % rel) : ((nextValue - rel) % rel == 0 || (nextValue - rel) % rel < 0) ? -rel : -((nextValue - rel) % rel)
```

Comment:
This line is a bit too much logic in one place imo, makes it a bit hard to read.

Reply:
Agree - expanded it out and added a comment.